src/plugins/data/public/search/session/session_service.ts

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { PublicContract } from '@kbn/utility-types';
+import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
+import { Observable, Subject, Subscription } from 'rxjs';
+import { PluginInitializerContext, StartServicesAccessor } from 'kibana/public';
+import { UrlGeneratorId, UrlGeneratorStateMapping } from '../../../../share/public/';
+import { ConfigSchema } from '../../../config';
+import { createSessionStateContainer, SessionState, SessionStateContainer } from './session_state';
+import { ISessionsClient } from './sessions_client';
+
+export type ISessionService = PublicContract<SessionService>;
+
+export interface TrackSearchDescriptor {
+  abort: () => void;
+}
+
+/**
+ * Provide info about current search session to be stored in backgroundSearch saved object
+ */
+export interface SearchSessionRestorationInfoProvider<ID extends UrlGeneratorId = UrlGeneratorId> {
+  /**
+   * User-facing name of the session.
+   * e.g. will be displayed in background sessions management list
+   */
+  getName: () => Promise<string>;
+  getUrlGeneratorData: () => Promise<{
+    urlGeneratorId: ID;
+    initialState: UrlGeneratorStateMapping[ID]['State'];
+    restoreState: UrlGeneratorStateMapping[ID]['State'];
+  }>;
+}
+
+/**
+ * Responsible for tracking a current search session. Supports only a single session at a time.
+ */
+export class SessionService implements ISessionService {
```

Comment:
Some of this is existing moved from `src/plugins/data/public/search/session_service.ts`

Session service was split into two services:

* SessionsClient (just SO CRUD)
* SessionSevice (all that is related to state management of current session)